bokehjs/src/lib/api/plotting.ts

```diff
@@ -630,30 +630,51 @@ export class Figure extends Plot {
     return this._marker(models.Scatter, args)
   }
 
-  _pop_colors_and_alpha(cls: Class<HasProps>, attrs: Attrs, prefix: string = "",
-                        default_color: Color = _default_color, default_alpha: number = _default_alpha): Attrs {
-    const result: Attrs = {}
+  _pop_visuals(cls: Class<HasProps>, props: Attrs, prefix: string = "",
+                        defaults: Attrs = {}, trait_defaults: Attrs = {}): Attrs {
 
-    const color = _with_default(attrs[prefix + "color"] as any, default_color)
-    const alpha = _with_default(attrs[prefix + "alpha"] as any, default_alpha)
+    const _split_feature_trait = function(ft: string): Array<string> {
+      const fta: Array<string> = ft.split('_', 2)
+      if (fta.length==2) { return fta }
+      else { return fta.concat(['']) }
+    }
+    const _is_visual = function(ft: string): boolean {
+      let feature, trait
+      [feature, trait] = _split_feature_trait(ft)
+      return includes(['line', 'fill', 'text', 'global'], feature) && (trait!=='')
+    }
 
-    delete attrs[prefix + "color"]
-    delete attrs[prefix + "alpha"]
+    defaults = {...defaults}
+    if ( !defaults.hasOwnProperty('text_color') ) {
+      defaults['text_color'] = 'black' }
+    trait_defaults = {...trait_defaults}
+    if ( !trait_defaults.hasOwnProperty('color') ) {
+      trait_defaults['color'] = _default_color }
+    if ( !trait_defaults.hasOwnProperty('alpha') ){
+      trait_defaults['alpha'] = _default_alpha }
 
-    const _update_with = function(name: string, default_value: unknown): void {
-      if (cls.prototype.props[name] != null) {
-        result[name] = _with_default(attrs[prefix + name], default_value)
-        delete attrs[prefix + name]
+    const result: Attrs = {}
+    const traits = new Set()
+    for (const pname in cls.prototype.props) {
+      if (_is_visual(pname)) {
+        let trait = _split_feature_trait(pname)[1]
+        if (props.hasOwnProperty(prefix+pname)) {
+            result[pname] = props[prefix+pname]
+            delete props[prefix+pname] }
+        else if (!cls.prototype.props.hasOwnProperty(trait)
+                 && props.hasOwnProperty(prefix+trait)) {
+            result[pname] = props[prefix+trait] }
+        else if (defaults.hasOwnProperty(pname)) {
+            result[pname] = defaults[pname] }
```

Comment:
Let's start from fixing this syntax (indentation, curly brackets, spacing, etc.), then we can get to fixing type errors.

Reply:
ok, i guess it doesn't hurt to install nodejs on my pc. linting fixes committed.